pkg/reconcile/pipeline/api.go

```diff
@@ -0,0 +1,175 @@
+package pipeline
+
+import (
+	"fmt"
+	olmalpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	sbapi "github.com/redhat-developer/service-binding-operator/api/v1alpha1"
+	"github.com/redhat-developer/service-binding-operator/pkg/binding"
+	"github.com/redhat-developer/service-binding-operator/pkg/client/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+//go:generate mockgen -destination=mocks/mocks_pipeline.go -package=mocks . Context,Service,CRD,Application,ContextProvider,Handler
+
+// Reconcile pipeline
+type Pipeline interface {
+
+	// Reconcile given service binding
+	// Returns true if processing should be repeated
+	// and optional error if occurred
+	// important: even if error occurred it might not be needed to retry processing
+	Process(binding *sbapi.ServiceBinding) (bool, error)
+}
+
+// A pipeline stage
+type Handler interface {
+	Handle(ctx Context)
+}
+
+// Pipeline flow control
+type FlowStatus struct {
+	Retry bool
+	Stop  bool
+	Err   error
+}
+
+// Service to be bound
+type Service interface {
+
+	// Service resource
+	Resource() *unstructured.Unstructured
+
+	// Return CRD for this service, otherwise nil if not backed by CRD
+	// Error might be returned if occurred during the operation
+	CustomResourceDefinition() (CRD, error)
+
+	// Resources owned by the service, if any
+	// Error might be returned if occurred during the operation
+	OwnedResources() ([]*unstructured.Unstructured, error)
+
+	// Attach binding definition to service
+	AddBindingDef(def binding.Definition)
+
+	// All binding definitions attached to the service
+	BindingDefs() []binding.Definition
+
+	// Optional service id
+	Id() *string
+}
+
+// Application to be bound to service(s)
+type Application interface {
+
+	// Application resource
+	Resource() *unstructured.Unstructured
+
+	// dot-separated path inside the application resource locating container resources
+	// the returns value follows foo.bar.bla convention
+	// it cannot be empty
```

Comment:
It defaults to a value but it is optional (Reference https://github.com/redhat-developer/service-binding-operator/blob/master/api/v1alpha1/servicebinding_types.go#L185)

Reply:
The contract on `pipeline.Application` is that it cannot be empty. This contact is different from the one you referred on ServiceBinding.